Enable more nullable reference types in asn.xslt.

vcsjones opened the pull request:
* string and Oid types on Choice can be null.
* Reference types that are `@optional` can be null.

Fixes #33824 

/cc @bartonjs

vcsjones committed:
Enable more nullable reference types in asn.xslt.

string and Oid types on Choice can be null.
Reference types that are @optional can be null.

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AlgorithmIdentifierAsn.xml.cs:
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
Most of these `#nullable enable` comes from the `asn.xslt`, this is generated code.

vcsjones commented on src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampRequest.cs:
```diff
@@ -38,7 +38,7 @@ public X509ExtensionCollection GetExtensions()
             }
 
-            X509ExtensionAsn[] rawExtensions = _parsedData.Extensions;
+            X509ExtensionAsn[] rawExtensions = _parsedData.Extensions!;
```
The `if (!HasExtensions)` on line 36 checks that `Extensions` is not null and > 0.

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs:
```diff
@@ -248,7 +248,7 @@ private void Decrypt(ReadOnlySpan<char> password, ReadOnlyMemory<byte> authSafeC
             // but it's sort of nonsensical when loading this way.
             CertBagAsn[] certBags = ArrayPool<CertBagAsn>.Shared.Rent(10);
-            AttributeAsn[][] certBagAttrs = ArrayPool<AttributeAsn[]>.Shared.Rent(10);
+            AttributeAsn[]?[] certBagAttrs = ArrayPool<AttributeAsn[]?>.Shared.Rent(10);
```
Line 688 could potentially assign null to an index in `certBagAttrs`

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/asn.xslt:
```diff
@@ -51,6 +51,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
Why is this necessary? Are these files included in a project that's not yet annotated?

vcsjones replied:
It may not be; one of the other XSLT templates in the same file had `#nullable enable`, so some of the files already had them:

https://github.com/dotnet/runtime/blob/419e949d258ecee4c40a460fb09c66d974229623/src/libraries/Common/src/System/Security/Cryptography/Asn1/asn.xslt#L130

Would it be preferable to get rid of all of them if possible now?

stephentoub replied:
If they're no longer necessary, yes, please.

vcsjones replied:
There are 5 projects that have `AsnXml` in them and they are have `<nullable>enable</nullable>`, so it appears that `#nullable enable` does not need to be in the template anymore.

stephentoub replied:
Cool.

vcsjones committed:
Remove #nullable enable.

All projects that include the ASN.1 generate files should already
have nullable enable project-wide.

stephentoub reviewed (approved):
Thanks

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixExportProvider.cs:
```diff
@@ -197,25 +197,26 @@ private byte[] ExportPfx(SafePasswordHandle password)
                 attrBytes[1] = sizeof(int);
                 MemoryMarshal.Write(attrBytes.AsSpan(2), ref keyIdx);
 
+                AttributeAsn attribute = new AttributeAsn
```
Why did this need to move out?

bartonjs replied:
Ah, presumably to avoid nullable annotations in the indirect reference it used to have.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixExportProvider.cs:
```diff
@@ -197,25 +197,26 @@ private byte[] ExportPfx(SafePasswordHandle password)
-                            }
-                        }
+                        attribute
                     }
```
Either

```C#
BagAttributes = new[] { attribute },
```

or


```C#
BagAttributes = new[]
{
    attribute,
},
```

, please.  (That's "Add the trailing comma, or collapse it to one line")

vcsjones committed:
Fix formatting from feedback.
